### PR TITLE
Improve Typst Lexer

### PR DESCRIPTION
**Pull request by Nerixyz**
This PR advances the Typst lexer. It's extended from https://github.com/pygments/pygments/pull/2656 by @jonaspleyer (see https://github.com/pygments/pygments/pull/2656#issuecomment-2156068478).

It adds a few snippets and most importantly, the example is now tokenized correctly. The example was fixed too (https://github.com/pygments/pygments/pull/2596#issuecomment-1858957697).

I described all commits I made in https://github.com/pygments/pygments/pull/2656#issuecomment-2156058113 (which mainly worked its way towards correctly handling the example).

~~It might be nice to add an option for the lexer to start in `code` (similar to the `startinline` option for PHP) - I'm not sure what the best implementation of that would be, though.~~
I've added the `start` option, which specifies the starting state of the lexer (one of 'markup', 'code', 'math') - this can be useful when highlighting Typst code inline.

---

**Original description of https://github.com/pygments/pygments/pull/2656:**

> This PR adds multiple improvements to the existing typst lexer and some tests
>
>  -  One of these tests is currently failing. This is by design.
> 
> I cannot guarantee that I will be working on this problem too much but wanted to get these commits out here.
Please tell me if the failing test prohibits merging. I can comment it out such that future development can reference it.
>
> Happy about any feedback.

### CODE REVIEW

**Comment by jonaspleyer** [comment]
Thanks @Nerixyz for your continued efforts on this topic!

File: pygments/lexers/typst.py [hunk 1]
```diff
@@ -27,49 +34,74 @@ class TypstLexer(RegexLexer):
...
+            (words(('#let', '#set', '#show'), suffix=r'\b'), Keyword.Declaration, 'inline_code'),
+            (words(('#import', '#include'), suffix=r'\b'), Keyword.Namespace, 'inline_code'),
+            (r'#\{', Punctuation, 'code'),
+            (r'#\(', Punctuation, 'code'),
+            (r'(#[a-zA-Z_][a-zA-Z0-9_-]*)(\[)', bygroups(Name.Function, Punctuation), 'markup'),
+            (r'(#[a-zA-Z_][a-zA-Z0-9_-]*)(\()', bygroups(Name.Function, Punctuation), 'code'),
+            (words(('#true', '#false', '#none', '#auto'), suffix=r'\b'), Keyword.Constant),
+            (r'#[a-zA-Z_][a-zA-Z0-9_]*', Name.Variable),
+            (r'#0x[0-9a-fA-F]+', Number.Hex),
+            (r'#0b[01]+', Number.Binary),
```
Comment (oliverpool) [inline pygments/lexers/typst.py:59]: Should be `Bin` according to https://github.com/pygments/pygments/blob/f64833d9d0443ff2443f8f8a4357c00196734d6c/pygments/token.py#L181

```suggestion
            (r'#0b[01]+', Number.Bin),
```

File: pygments/lexers/typst.py [hunk 2]
```diff
@@ -81,24 +113,47 @@ class TypstLexer(RegexLexer):
...
+            (r'([a-zA-Z_][a-zA-Z0-9_-]*)(\()', bygroups(Name.Function, Punctuation), 'code'),
             (words(('as', 'break', 'export', 'continue', 'else', 'for', 'if',
-                    'import', 'in', 'include', 'return', 'while'), suffix=r'\b'),
+                    'in', 'return', 'while'), suffix=r'\b'),
              Keyword.Reserved),
+             (words(('import', 'include'), suffix=r'\b'), Keyword.Namespace),
             (words(('auto', 'none', 'true', 'false'), suffix=r'\b'), Keyword.Constant),
             (r'([0-9.]+)(mm|pt|cm|in|em|fr|%)', bygroups(Number, Keyword.Reserved)),
+            (r'0x[0-9a-fA-F]+', Number.Hex),
+            (r'0b[01]+', Number.Binary),
```
Comment (oliverpool) [inline pygments/lexers/typst.py:129]: Should be `Bin` according to https://github.com/pygments/pygments/blob/f64833d9d0443ff2443f8f8a4357c00196734d6c/pygments/token.py#L181

```suggestion
            (r'#0b[01]+', Number.Bin),
```

**Review by oliverpool — COMMENTED** [review]
Thanks for the work, it looks good!

(note: not a maintainer, just a random user)

**Comment by Nerixyz** [comment]
> Thanks for the work, it looks good!
> 
> (note: not a maintainer, just a random user)

Oopsie, I should've read the [docs](https://pygments.org/docs/tokens/#literals) better. Thank you for pointing that out!

**Comment by oliverpool** [comment]
That was quick :smile: 

I imported your lexer into [chroma](https://github.com/alecthomas/chroma/issues/877) and tested it with the following snippet:

```typst
#let song(
  title: none,
  title_index: none,
  singer: none,
  singer_index: none,
  references: (),
  line_color: rgb(0xd0, 0xd0, 0xd0),
  header_display: (number, title, singer) => grid(
    columns: (auto,1fr),
    inset: 4pt,
    align: horizon,
    grid.cell(fill: rgb(0xd0, 0xd0, 0xd0),align: top, inset:(rest:4pt, /*right: 10pt*/), heading(outlined: false,str(number))),
    [
      #heading(title)
      #if singer != none {
        strong(singer) + "\n"
      }
    ],
  ),
  // comment
  doc,
)
```

the `\n` looks off (and the nearby `#if` is not styled):

image


This may be due to the conversion to chroma's xml (I don't know how to use pygment to quickly render a snippet, so maybe this issue is not present here), but maybe you can reproduce it/test it as well?

**Comment by Nerixyz** [comment]
> the `\n` looks off (and the nearby `#if` is not styled):

I missed that. Should be supported now (although with some quirks):

```typ
#if foo {
  bar
} else {
  baz
} qox
```

In this case, `qox` would become a variable because the "inline_code" context lasts until the next newline or `;`

File: scripts/debug_lexer.py [hunk 3]
```diff
@@ -162,24 +163,24 @@ def main(fn, lexer=None, options={}):
...
+        if lxcls.__bases__ == (RegexLexer,):
+            lxcls.__bases__ = (DebuggingRegexLexer,)
+            debug_lexer = True
+        elif lxcls.__bases__ == (DebuggingRegexLexer,):
+            # already debugged before
+            debug_lexer = True
+        else:
+            # HACK: ExtendedRegexLexer subclasses will only partially work here.
+            lxcls.__bases__ = (DebuggingRegexLexer,)
+            debug_lexer = True
```
Comment (jvoisin) [inline scripts/debug_lexer.py:183]: This looks unrelated to the PR :)

**Review by jvoisin — APPROVED** [review]

**Review by oliverpool — APPROVED** [review]
With the latest changes, my own typst examples look very good :+1: (I don't use math, but a good mix of markup, content and functions).

**Comment by Nerixyz** [comment]
> With the latest changes, my own typst examples look very good 👍 (I don't use math, but a good mix of markup, content and functions).

Thank you for providing feedback! I also use it on https://nerixyz.github.io/icu-typ/ (mostly code). If there are any further issues, feel free to @ me.

**Comment by Anteru** [comment]
I'll trust you guys you know what you're doing here, the lexing looks fine, but I'll clean it up a bit to merge successive punctuation tokens.
